src/main/java/org/openrewrite/java/dependencies/ChangeDependency.java

```diff
@@ -82,31 +87,24 @@ public String getDescription() {
         return "Change the groupId, artifactId and/or the version of a specified Gradle or Maven dependency.";
     }
 
-    private final org.openrewrite.gradle.ChangeDependency changeGradleDependency;
-    private final org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId changeMavenDependency;
-
-    public ChangeDependency(
-            String oldGroupId,
-            String oldArtifactId,
-            @Nullable String newGroupId,
-            @Nullable String newArtifactId,
-            @Nullable String newVersion,
-            @Nullable String versionPattern,
-            @Nullable Boolean overrideManagedVersion
-    ) {
-        this.oldGroupId = oldGroupId;
-        this.oldArtifactId = oldArtifactId;
-        this.newGroupId = newGroupId;
-        this.newArtifactId = newArtifactId;
-        this.newVersion = newVersion;
-        this.versionPattern = versionPattern;
-        this.overrideManagedVersion = overrideManagedVersion;
-        changeGradleDependency = new org.openrewrite.gradle.ChangeDependency(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern);
-        changeMavenDependency = new org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern, overrideManagedVersion);
-    }
+    @Nullable
+    private transient org.openrewrite.gradle.ChangeDependency changeGradleDependency;
+    @Nullable
+    private transient org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId changeMavenDependency;
 
     @Override
     public List<Recipe> getRecipeList() {
+        if (changeGradleDependency == null ||
+                !Objects.equals(changeGradleDependency.getOldGroupId(), oldGroupId) ||
+                !Objects.equals(changeGradleDependency.getOldArtifactId(), oldArtifactId) ||
+                !Objects.equals(changeGradleDependency.getNewGroupId(), newGroupId) ||
+                !Objects.equals(changeGradleDependency.getNewArtifactId(), newArtifactId) ||
+                !Objects.equals(changeGradleDependency.getNewVersion(), newVersion) ||
+                !Objects.equals(changeGradleDependency.getVersionPattern(), versionPattern)
+        ) {
+            changeGradleDependency = new org.openrewrite.gradle.ChangeDependency(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern);
+            changeMavenDependency = new org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern, overrideManagedVersion);
+        }
         return Arrays.asList(
                 changeGradleDependency,
                 changeMavenDependency
```

Comment:
Aren't we over complicating things here with the early instantiation of those recipes?
```suggestion
    @Override
    public List<Recipe> getRecipeList() {
        return Arrays.asList(
                new org.openrewrite.gradle.ChangeDependency(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern),
                new org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern, overrideManagedVersion)
```

The only question is then how that translates to potentially repeated calls to `getRecipeList()`, but I'd _think_ that's manageable for the increased simplicity we get here.

Reply:
it gets called at least 3 times every cycle (scan + generate + edit) + 1 validateAll at the beginning + 1 getRecipeDescriptor...

Reply:
What about just a simple hybrid?

```suggestion
@Override
    public List<Recipe> getRecipeList() {
        if (changeGradleDependency == null) {
            changeGradleDependency = new org.openrewrite.gradle.ChangeDependency(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern);
            changeMavenDependency = new org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId(oldGroupId, oldArtifactId, newGroupId, newArtifactId, newVersion, versionPattern, overrideManagedVersion);
        }
        return Arrays.asList(changeGradleDependency, changeMavenDependency);
```

I would think that given we initialize with the options as correctly populated then we should already know that we've set their options to what we have per the current outer recipe definition. Do we need to recheck the sub-recipes options every pass?

Reply:
The problem is that the recipes are originally instantiated via [RecipeInstrospectionUtils]( https://github.com/openrewrite/rewrite/blob/084fbd8b28e0724d68e837de6baa3d7d96445851/rewrite-core/src/main/java/org/openrewrite/internal/RecipeIntrospectionUtils.java), which populates them with default values and nulls, then getRecipeList can be called to get the recipeDescriptors, and later on the same recipe (or a clone of it) is getting it's parameters populated via jackson reflection, which can cause the children recipes to have different values than the parent one if not instantiated in the right moment... That's why checking always if the parameters are different seems the safest approach...